src/cargo/ops/cargo_compile.rs

```diff
@@ -116,12 +116,19 @@ pub enum CompileFilter<'a> {
 
 pub fn compile<'a>(ws: &Workspace<'a>, options: &CompileOptions<'a>)
                    -> CargoResult<ops::Compilation<'a>> {
+    compile_with_exec(ws, options, &mut DefaultExecutor)
+}
+
+pub fn compile_with_exec<'a, E: Executor>(ws: &Workspace<'a>,
+                                          options: &CompileOptions<'a>, 
+                                          exec: &mut E)
```

Comment:
This shouldn't affect Cargo's compile times at all - there is only one actual type which implements Executor, so unless you're using Cargo as a lib, exactly the same amount of code is processed as before. The only downside is we add a generic parameter in a few places, which is a bit ugly, but is also kind of best practice (IMO). Speed is important here, but you are right in that the speed improvement (vs trait object) is unlikely to be important.